react/ComponentStyle.md

```diff
@@ -0,0 +1,304 @@
+  * if you want to use props or both props and context in your constructor, you need to pass them to the constructor of its super class.
+
+    ```
+    constructor(props) {
+      super(props);
+      // this.props....
+    }
+
+    constructor(props, context) {
+      super(props, context);
+      // this.props, this.context...
+    }
+    ```
+  * bind event-handler in constructor.
+
+* styles
+  * constructor: follow this order: super -> set initial state -> bind event-handler
+    ```
+    constructor() {
+      super();
+      this.state = {};
+      this.handleSomethingClick = this.handleSomethingClick.bind(this);
+    }
+    ```
+  * event-handler: if the handler is used by DOMElement named **button**, the corresponding event is **Click**, you should write it as
+
+    ```
+    constructor() {
+      ...
+      this.handleButtonClick = this.handleButtonClick.bind(this);
+    }
+
+    handleButtonClick() {
+
+    }
+
+    render() {
+      return <div className={styles.button} onClick={this.handleButtonClick}></div>
+    }
+    ```
+
+    that is to say, the event-handler should be named **handleElementEvent**
+  * order of Component methods should be as follows:
+    [constructor] -> [lifecycle methods] -> [self-defined methods] -> render
+  * if render method is to long, and it's a combination of several parts. Extract each part as a single method. For instance:
+
+    ```
+    class Madadata extends PureComponent {
+      render() {
+        return (
+          <div className={styles.container}>
+            <div className={styles.header}>
+            </div>
+            <div className={styles.body}>
+            </div>
+            <div className={styles.footer}>
+            </div>
+          </div>
+        );
+      }
+    }
+    ```
+
+    you'd better extract header, body and footer out like:
+
+    ```
+    class Madadata extends PureComponent {
+
+      getHeaderJSX() { // or you can name it **renderHeader**
+        return <div className={styles.header}></div>;
+      }
+
+      getBodyJSX() { // or you can name it **renderBody**
+        return <div className={styles.body}></div>;
+      }
+
+      getFooterJSX() { // or you can name it **renderFooter**
+        return <div className={styles.footer}></div>;
+      }
+
+      render() {
+        const headerJSX = this.getHeaderJSX();
+        const bodyJSX = this.getBodyJSX();
+        const footerJSX = this.getFooterJSX();
+        return (
+          <div className={styles.container}>
+            {headerJSX}
+            {bodyJSX}
+            {footerJSX}
+          </div>
+        );
+      }
+    }
+    ```
+
+    and these extracted methods should be right above render method.
+
+# Component
+* when to use?
+  * has lifecycle method
+  * has state to keep.
+  * UI is not pure component.
+
+* [others] same with PureComponent.
+
+# Redux
+
+* when to use?
+
+* styles
+  * prefer arrow functions for mapStateToProps and mapDispatchToProps:
```

Comment:
just convention. If you think that is not appropriate, we can use other conventions. My motivation for that is just for consistency.

this is good
```
function mapStateToProps() {}
function mapDispatchToProps() {}
```

this is good, as well as concise and neat.
```
const mapStateToProps = state => ({});
const mapDispatchToProps = ({});
```